osu.Game.Rulesets.Osu/Difficulty/MathUtil/FittsLaw.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+
+using MathNet.Numerics;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.MathUtil
+{
+    public static class FittsLaw
+    {
+        /// <summary>
+        /// Calculates the index of performance for the distance and the movement time specified.
+        /// Index of performance is the difficulty of a movement.
+        /// </summary>
+        public static double CalculateIP(double d, double mt)
+        {
+            return Math.Log(d + 1, 2) / (mt + 1e-10);
+        }
+
+        /// <summary>
+        /// Calculates the probability that the target is hit successfully.
+        /// </summary>
+        public static double CalculateHitProb(double d, double mt, double tp)
+        {
+            if (d == 0)
+                return 1.0;
+
+            if (mt * tp > 50)
+                return 1.0;
+
+            if (mt <= 0.03)
+                mt = 0.03;
+
+            return SpecialFunctions.Erf(2.066 / d * (Power2(mt * tp) - 1) / Math.Sqrt(2));
+        }
+
+        private static readonly double[] coeffs = {
+            1.0000000060371126,
+            0.693146840098149,
+            0.2402310826131064,
+            0.05547894683131716,
+            0.009686150703032881,
+            0.0012382531241478965,
+            0.00021871427263121524,
+        };
+
+        /// <summary>
+        /// Fast approximation of 2^x. Accurate to around 9-10 significant figures, around 6x faster than Math.Pow or Math.Exp
+        /// Calculates the integer part using a bit shift and fraction part using a polynomial approximation
+        /// </summary>
+        public static double Power2(double x)
+        {
+            if (x<0)
+            {
+                return 1 / Power2(-x);
+            }
+            if (x>60)
+            {
+                return double.PositiveInfinity;
+            }
+
+            int floor = (int)x;
+            double frac = x - floor;
```

Comment:
This is a bit difficult to read. In theory a loop could make this much clearer while making it easier to add/remove coefficients without modifying this code:
```
int floor = (int)x;
double coeffSum= coeffs[0];
double frac = x - floor;
for(int i= 1; i < coeffs.Count(); i++) {
        coeffSum+= coeffs[i] * frac;
        frac *= frac;
}
return (1L << floor) * coeffSum;
```
or something along those lines